### PR TITLE
Punch up the Racer's Url Handling in its server

### PR DESCRIPTION
**Pull request by spember**
-created a basic 'endpoint' object and approach to handling urls with…
-shifted the fetch and fingerprint endpoints into a dedicated directory

## Description

creates a bit more structure in the racer's server and url handling. doing this before I add the lighthouse running

### CODE REVIEW

File: packages/racer/src/controllers/fingerprint.ts [hunk 1]
```diff
@@ -6,26 +6,25 @@
...
 
   const response = await fetch(FINGERPRINT_ENDPOINT, {
-    agent: httpsAgent,
+    agent: RaceProxyHttpsAgent,
   });
   const data = await response.json();
   const spkiFingerprint = data?.spkiFingerprint;
   console.log(`Received fingerprint: ${spkiFingerprint}`);
-  return spkiFingerprint;
+  return {fingerprint: spkiFingerprint};
```
Comment (JasonPapanicholasWF) [inline packages/racer/src/controllers/fingerprint.ts]: Do we just want to change it to `fingerprint` in both places?

**Comment by spember** [comment]
to try it out yourself:

```
yarn build && docker-compose up --build
```

then from your host machine:

`curl localhost:3000/fingerprint`
`curl "localhost:3000/fetch?size=1000&url=https://www.google.com"`

File: packages/racer/src/server.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import http from 'http';
+import {IncomingMessage, ServerResponse} from 'http';
+import {RegisteredEndpoint} from './controllers/common';
+import {FetchEndpoint} from './controllers/fetch';
+import {FingerprintEndpoint} from './controllers/fingerprint';
+
+const endpoints: RegisteredEndpoint[] = [FetchEndpoint, FingerprintEndpoint];
+
+export const initialize = (): http.Server => {
+  return http.createServer((req: IncomingMessage, res: ServerResponse) => {
```
Comment (JasonPapanicholasWF) [inline packages/racer/src/server.ts:11]: I'm a little confused as to what the incoming message request URL looks like in this. Is it `{proxyIP}/fetch?url={realDestination}` or in the case of the fingerprint `{proxyIP}/fingerprint` ?

File: packages/racer/src/index.ts [hunk 3]
```diff
@@ -1,34 +1,10 @@
...
-      } else {
-        res.statusCode = 404;
-      }
-
-      res.end();
-    }
-  );
+const goSpeedRacer = async () => {
+  const server = initialize();
   console.log('Go Speed Racer Go');
```
Comment (JasonPapanicholasWF) [inline packages/racer/src/index.ts:7]: 🏎️  🐵 🥷 🎇

File: packages/racer/src/controllers/common.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+export interface RegisteredEndpoint {
+  path: string;
+  handler: (
+    request: IncomingMessage,
+    response: ServerResponse,
+    url: url.UrlWithParsedQuery
+  ) => Promise<any>;
+}
+
+export const RaceProxyHttpsAgent = new https.Agent({
```
Comment (JasonPapanicholasWF) [inline packages/racer/src/controllers/common.ts:15]: Do we need to support http here?

Comment (spember) [reply packages/racer/src/controllers/common.ts:15]: probably, but I didn't have a need for it yet

File: packages/racer/src/server.ts (hunk 2 above)
Comment (spember) [reply packages/racer/src/server.ts:11]: yeah it takes a second or two to wrap your head around it. This server is the `racer`'s server, meaning that each racer node will be able to accept http requests from the outside (e.g. your laptop).

So, once you do docker compose up, you could then (from your terminal, not within a docker container) do something like `curl localhost:3000/fingerprint`: the racer receives this request, checks the path, sees if it has a matching Endpoint (it does), then _the racer_ turns around and talks to the proxy, then sends the response from the proxy back to you.

Does that help? see my example url above for triggering a fetch. in that case, the racer receives the request from you, then attempts to call the url you provide in `url`... _thinking it's calling the real url_, but it is, in reality, hitting the proxy

File: packages/racer/src/controllers/common.ts (hunk 4 above)
Comment (spember) [reply packages/racer/src/controllers/common.ts:15]: actually wait, I'll add it

**Review by JasonPapanicholasWF — APPROVED** [review]

**Pull request merged**